sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -222,6 +222,20 @@ abstract class Expression extends TreeNode[Expression] {
    */
   def childrenResolved: Boolean = children.forall(_.resolved)
 
+  // Return a roughly canonicalized expression, which recursively canonicalizes each expression node
```

Comment:
Maybe it's best to keep these as regular Scaladoc `/** ... */ ` instead of comments.